src/machine/machine_atmega328p.go

```diff
@@ -88,3 +89,132 @@ func (pwm PWM) Set(value uint16) {
 		panic("Invalid PWM pin")
 	}
 }
+
+const (
+	SPI_CLOCK_FOSC4         uint8 = 0
+	SPI_CLOCK_FOSC16        uint8 = 1
+	SPI_CLOCK_FOSC64        uint8 = 2
+	SPI_CLOCK_FOSC128       uint8 = 3
+	SPI_CLOCK_FOSC2         uint8 = 4
+	SPI_CLOCK_FOSC8         uint8 = 5
+	SPI_CLOCK_FOSC32        uint8 = 6
+	SPI_CLOCK_FOSC64_Double uint8 = 7
+	SPI_CLOCK_MASK          uint8 = 0x03
+	SPI_2XCLOCK_MASK        uint8 = 4
+)
+
+// SPIConfig
+type SPIConfig struct {
+	LSBfirst bool
+	// Frequency defaults to 6000000Hertz if not set
+	Frequency uint32
+	// Mode can be 0, 1, 2, 3, 4, 5
+	Mode uint8
+	SDI  Pin
+	SDO  Pin
+	SCK  Pin
+}
+
+// SPI is for the Serial Peripheral Interface
+// Data is taken from http://ww1.microchip.com/downloads/en/DeviceDoc/ATmega48A-PA-88A-PA-168A-PA-328-P-DS-DS40002061A.pdf page 169 and following
+type SPI struct {
+}
+
+// SPI0 is the default SPI interface
+var SPI0 = SPI{}
+
+// Configure uses the given config to setup the SPI interface
+func (spi SPI) Configure(config SPIConfig) error {
+	// Use default pins if not set.
+	if config.SCK == 0 && config.SDO == 0 && config.SDI == 0 {
+		config.SCK = PB5
+		config.SDO = PB3
+		config.SDI = PB4
+	}
+
+	// set default frequency
+	if config.Frequency == 0 {
+		config.Frequency = 6000000
+	}
+
+	var clockDivider uint8
+	switch {
+	case config.Frequency >= 93750 && config.Frequency < 187500:
+		clockDivider = SPI_CLOCK_FOSC128
```

Comment:
2. Does this really make any sense here? We do not set a specific frequency in the SPI Controller Register, but we set a clock divider. So i guess, even if a custom clock rate is being used it will still use the clock divider on that custom clock rate. 

But i am really not an expert for the atmega328p. That is just how i understand the documentation.

Reply:
Okay, i got it now, yes i'll try to consider that. Taking a nap sometimes helps understanding things :)